packages/node/src/anr/index.ts

```diff
@@ -0,0 +1,182 @@
+   */
+  pollInterval: number;
+  /**
+   * Threshold in milliseconds to trigger an ANR event.
+   *
+   * Defaults to 5000ms.
+   */
+  anrThreshold: number;
+  /**
+   * Whether to capture a stack trace when the ANR event is triggered.
+   *
+   * Defaults to `false`.
+   *
+   * This uses the node debugger which enables the inspector API and opens the required ports.
+   */
+  captureStackTrace: boolean;
+  /**
+   * Log debug information.
+   */
+  debug: boolean;
+}
+
+function sendEvent(blockedMs: number, frames?: StackFrame[]): void {
+  const event: Event = {
+    level: 'error',
+    exception: {
+      values: [
+        {
+          type: 'ApplicationNotResponding',
+          value: `Application Not Responding for at least ${blockedMs} ms`,
+          stacktrace: { frames },
+          mechanism: {
+            // This ensures the UI doesn't say 'Crashed in' for the stack trace
+            type: 'ANR',
+          },
+        },
+      ],
+    },
+  };
+
+  captureEvent(event);
+
+  void flush(3000).then(() => {
+    // We only capture one event to avoid spamming users with errors
+    process.exit();
+  });
+}
+
+function startChildProcess(options: Options): void {
+  const env = { ...process.env };
+
+  if (options.captureStackTrace) {
+    inspector.open();
+    env.SENTRY_INSPECT_URL = inspector.url();
+  }
+
+  const child = fork(options.entryScript, {
+    env,
+    stdio: options.debug ? 'inherit' : 'ignore',
+  });
+  // The child process should not keep the main process alive
+  child.unref();
+
+  const timer = setInterval(() => {
+    try {
+      // message the child process to tell it the main event loop is still running
+      child.send('ping');
+    } catch (_) {
+      //
+    }
+  }, options.pollInterval);
+
+  child.on('error', () => {
+    clearTimeout(timer);
+  });
+  child.on('disconnect', () => {
+    clearTimeout(timer);
+  });
+  child.on('exit', () => {
+    clearTimeout(timer);
+  });
+}
+
+function handleChildProcess(options: Options): void {
+  addGlobalEventProcessor(event => {
+    // Strip sdkProcessingMetadata from all child process events to remove trace info
+    delete event.sdkProcessingMetadata;
+    return event;
+  });
+
+  let debuggerPause: Promise<() => void> | undefined;
+
+  // if attachStackTrace is enabled, we'll have a debugger url to connect to
+  if (process.env.SENTRY_INSPECT_URL) {
+    debuggerPause = captureStackTrace(process.env.SENTRY_INSPECT_URL, frames => {
+      sendEvent(options.anrThreshold, frames);
+    });
+  }
+
+  async function watchdogTimeout(): Promise<void> {
+    const pauseAndCapture = await debuggerPause;
+
+    if (pauseAndCapture) {
+      pauseAndCapture();
+    } else {
+      sendEvent(options.anrThreshold);
+    }
+  }
+
+  const ping = watchdogTimer(options.pollInterval, options.anrThreshold, watchdogTimeout);
+
+  process.on('message', () => {
+    ping();
+  });
+}
+
+/**
+ * Starts a child process that detects Application Not Responding (ANR) errors.
```

Comment:
Let's add a note that this is going to be in beta state so we have the liberty to make breaking changes if absolutely necessary.